---
title: "FOR-271 RNAseq project"
author:
  - name: Jake Payne, 018465886
output: html_document
---

In this project I will be using RStudio (v.4.4) to analyse RNAseq results, following the instructions available at https://zilanwen.github.io/RNA-Seq-Differential_Gene_Expression/.

## Part 1: Preliminaries

Loading the required libraries and setting up the data structure.

```{r echo=T, message=FALSE, error=FALSE}
library(edgeR)
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(rmarkdown)
library(knitr)
```

It's important to examine code and files before and during work, and I am grateful there are numerous "checks" included in the instructions, but for cleaner presentation here, I've commented most of them out.

Data:
```{r}

setwd('C:/Users/Asus/Documents/GitHub/RNA_seq_DEA_Jake_P')
Coinfection.targets<-read.delim("C:/Users/Asus/Documents/GitHub/RNA_seq_DEA_Jake_P/data/fileDesc.txt")
rownames(Coinfection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
Coinfection.orig <- readDGE(Coinfection.targets, header=F)
#head(Coinfection.orig)
Coinfection.rawCount <- Coinfection.orig$count
#dim(Coinfection.rawCount)
#head(Coinfection.rawCount)
```

Metadata:
```{r}
#setting up metadata
Treatment <- factor(c(rep("Ha",3), rep("Ctr", 3)))
meta <- data.frame(Treatment, row.names = colnames(Coinfection.orig$count))
#colnames(Coinfection.orig$count)
#rownames(meta)
all(colnames(Coinfection.orig$count) %in% rownames(meta)) & all(rownames(meta) %in% colnames(Coinfection.orig$count)) 
#^ A last final check to make sure row & column names match exactly
```
Setting up size factors and normalizing counts:
```{r message =FALSE}
dds <- DESeqDataSetFromMatrix(Coinfection.orig, colData = meta, design = ~ Treatment)
#head(counts(dds))
dds <- estimateSizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
write.csv(normalized_counts, file="./results/coinfection_normalized_counts_DESeq2.csv")
sizeFactors(dds) #the size factors each variable is normalized by:
head(counts(dds)) #non normalized counts:
head(normalized_counts) #normalized counts:
```

## Part 2: Performing QC

### Principal Component Analysis

```{r}
rld <- rlog(dds, blind=TRUE)
plotPCA(rld, intgroup="Treatment")
pdf("./results/PlotPCA_dds.pdf")
png("./results/PlotPCA_dds.png") #I'm not clear why the instructions say to save this as PDF, so I've also included it as a png file.
dev.off()
```

From the above plot we can see that:

- 1. In PC1 (which presumably corresponds to infection status), the two groups cluster strongly with one another- there is much more difference between groups than within them (and this PC alone explains 50% of variance).PC2 is strongly affected by an outlier in the control group- without it it probably would not be the second principal component. We could remove the outlier here, but otherwise we can probably ignore PC2 as if the outlier were removed all other differences would be negligible.

- 2. PC1 very much fits the experimental expectation. It shows strong between-group differences, so we know that the treatment has an effect. PC2 is probably just a wildcard and does not affect our expectations.

- 3. The axis labels show the amount of variance explained by each component (i.e., 50% for PC1, 25% for PC2)

### Hierarchical clustering heatmap   

```{r}
rld_mat <- assay(rld) #convert to a form usable by pheatmap()
rld_cor <- cor(rld_mat) #pairwise correlations
head(rld_cor)

head(meta)

pheatmap(rld_cor, annotation = meta)
pdf("./results/PlotHeatmap_dds.pdf")

```

Changing the heatmap's colours and borders to make it more readable:

```{r}
heat.colors <- RColorBrewer::brewer.pal(6, "Blues")
pheatmap(rld_cor, annotation = meta, color = heat.colors, border_color=NA, fontsize = 10, 
         fontsize_row = 10, height=20)
```

So again, we can see quite strong clustering between the three Ha samples, and between the first two controls (with Ctr3 again being an outlier). I really like the hierarchical tree that the graph displays along with the heatmap.


### Differential Expression Analysis


```{r}
#getwd()
#setwd("C:/Users/username/Desktop/RNA-seq_DEA")
options(digits=3)
infection.targets<-read.delim("./data/fileDesc.txt")
infection.targets
rownames(infection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
infection.targets
infection <- readDGE(infection.targets, header=F)
#dim(infection)
#head(infection)
infection.rawCount <- infection$count
#head(infection.rawCount)
write.csv(infection.rawCount, file="./results/infection.rawCounts.csv") #moved this line as it makes more sense here IMO
```

Making a graph of Ha1's raw (unfiltered) counts (and exporting it):
```{r}
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
png("./results/count distribution.png", res=300, height=1800, width=1800)
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
dev.off()
```

CPM before filtering:
```{r}
infection.normCPM <- cpm(calcNormFactors(infection)) #counts per mil
dim(infection.normCPM)
head(infection.normCPM)
write.csv(infection.normCPM, file="./results/infection.normCPM.csv")
```

Filtering the infections data- expression must be in min. 1 CPM in min 3 libraries
```{r}
infection.filtered <- rowSums(cpm(infection)>1) >=3

table(infection.filtered)
#infection$samples$lib.size
Infection <- infection[infection.filtered,] 
#colSums(Infection$counts)
#dim(Infection)
Infection$samples$lib.size <- colSums(Infection$counts)
Infection$samples
Infection = calcNormFactors(Infection)
Infection$samples
Infection.filtered.normCPM <-cpm(calcNormFactors(Infection))
write.csv(Infection.filtered.normCPM, file="./results/Infection.filtered.normCPM.csv")
```

```{r}
group<-factor(c('Ha','Ha','Ha',"Ctr","Ctr","Ctr"))
Infection.design <- model.matrix(~group)   
rownames(Infection.design)<-colnames(Infection$counts)
Infection.design
plotMDS(Infection, main="MDS plot of RNA-Seq", labels=colnames(Infection$counts))
png("./results/plotMDS.Infection.png", res=300, height=1800, width=1800)
plotMDS(Infection, main="MDS plot of Infection RNA-Seq", labels=colnames(Infection$counts))
dev.off()
```
MDS versus PCA- MDS focuses on preserving pairwise distances, while PCA focuses on capturing maximum variance in the data.For practical purposes if we compare the two graphs:

[](./results/PlotPCA_dds.png)

We can see that MDS has variance on only the Y axis, with *distance* on X. PCA has variance on both axes. 

```{r}
Infection <- estimateGLMCommonDisp(Infection, Infection.design)
Infection <- estimateGLMTrendedDisp(Infection, Infection.design)
Infection <- estimateGLMTagwiseDisp(Infection, Infection.design)
```

```{ warning = F}
plotMeanVar(Infection, show.tagwise.vars=T,NBline=T)
plotBCV(Infection)
Infection.fit <- glmFit(Infection, Infection.design)
colnames(Infection.fit)
lrt.Ha_vs_Ctr <- glmLRT(Infection.fit, coef=2)  # to compare Ha vs Ctr (Ha_vs_Ctr)
t1<-topTags(lrt.Ha_vs_Ctr, n=nrow(Infection))

summary(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))
head(t1$table)
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC > 0))
lrt.Ha_vs_Ctr_UP <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC > 0)
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC < 0))
lrt.Ha_vs_Ctr_DW <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC < 0)

DEtags.lrt.Ha_vs_Ctr <- rownames(Infection)[as.logical(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))]

write.csv(lrt.Ha_vs_Ctr_UP, file="./results/lrt.Ha_vs_Ctr_UP.csv")
write.csv(lrt.Ha_vs_Ctr_DW, file="./results/lrt.Ha_vs_Ctr_DW.csv")

Infection.colHavsCtr = rep('grey55', nrow(Infection))

Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC >0 ] <- "red"
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC <0 ] <- "blue"

par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, de.tags=DEtags.lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", pch=19, cex=0.4, smearWidth=0.5, panel.first=grid(), smooth.scatter=FALSE, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
```

In an alternate colour:

```{r}
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="a infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
```

Final export!
```{r}
png("./results/plotSmear.InfectionRNAseq.png", res=300, height=1800, width=1800)
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
dev.off()
```







